flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java

```diff
@@ -1010,6 +1011,56 @@ public void testConcurrentRecycleAndRelease2() throws Exception {
 		}
 	}
 
+	@Test
+	public void testConcurrentGetNextBufferAndRelease() throws Exception {
```

Comment:
I think we could have tested this bug without concurrency/multi threading by just calling `releaseAllResources` before `getNextBuffer`. Test would be much easier to understand and debug and this would be worth a bit worse testing coverage - especially that we still have ITCases, and in the long run we are planning/hoping to make releasing resources go through mailbox.

Reply:
I am not quite sure whether ITCase can stable reproduce the concurrent issues if not constructed well. 

E.g. we need to guarantee that when the canceler thread is releasing the channel, this channel must already be queued into the input gate with data in `receivedBuffers`, then it can construct the scenario of executing `releaseAllResources` and `getNextBuffers` concurrently.

Unit test can work on these two methods directly to stable reproduce the potential bugs, but ITCase might be hard to control and it might never have the chance to enter this path in practice.

I found most of the previous race condition bugs in core codes was just missing such concurrent unit tests before, and simply verify the results by executing related methods by sequence, not concurrently. 

E.g. If we execute the #releaseAllResources() in different steps of `#getNextBuffer`, it would have different effects. So I think only unit test can cover all the potential possibilities if we are not sure ITCase can achieve it.

        ```
Optional<BufferAndAvailability> getNextBuffer() throws IOException {
                 **// 1. execute #releaseAllResources() here**
  
		checkState(!isReleased.get(), "Queried for a buffer after channel has been closed.");
		checkState(partitionRequestClient != null, "Queried for a buffer before requesting a queue.");

		checkError();

		final Buffer next;
		final boolean moreAvailable;

                 **//2. execute #releaseAllResources() here**

		synchronized (receivedBuffers) {
			next = receivedBuffers.poll();
			moreAvailable = !receivedBuffers.isEmpty();
		}

                 **//3. execute #releaseAllResources() here**

		numBytesIn.inc(next.getSize());
		numBuffersIn.inc();
		return Optional.of(new BufferAndAvailability(next, moreAvailable, 0));
	}
```

Maybe it seems complex to let unit tests handle such scenarios. If ITCase can handle this work well, i am happy to make only simple unit tests.

Reply:
I'm not a fan of such concurrency tests, I wouldn't be personally adding them I think, but you are right - single threaded unit test wouldn't reproduce this issue because of `checkState(!isReleased.get())`.

Reply:
I thought of some other considerations for this issue to share.

In the ITCase, even though we can reproduce some potential concurrent bugs, it is hard to debug and find the root cause, because it is involved in all the components. I really have such feeling when debugging the `UnalignedCheckpointITCase` these days.

Reversely, unit test only works on two concurrent methods directly, so it is easy to find the bugs by limiting the scopes/components. We already had 6 unit tests written by concurrent way in `RemoteInputChannelTest` before, to guarantee the stability among different concurrent methods executed by task thread, netty thread, canceler thread separately.  If replaced by ITCase, we need to debug among all these methods to find the potential root cause.

In general, it is better for unit tests only focus on one component or less, otherwise we should rely on ITCase. In this case, we only limit the scope inside `RemoteInputChannel` component, so it also makes sense from this aspect. 

Anyway besides the pros I mentioned above for unit tests, I also agree that the cons you concerned, merely the pros are a bit more than cons on my side. :)